File: src/com/ptby/dynamicreturntypeplugin/index/ReturnInitialisedSignatureConverter.java
package com.ptby.dynamicreturntypeplugin.index;

import com.intellij.openapi.project.Project;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;
import com.jetbrains.php.lang.psi.elements.impl.FunctionImpl;

import java.util.Collection;

public class ReturnInitialisedSignatureConverter {


    public ReturnInitialisedSignatureConverter() {
    }


    public String convertSignatureToClassSignature( String signature, Project project ) {
        String[] split = signature.split( ":" );
        PhpIndex phpIndex = PhpIndex.getInstance( project );

        if( split.length < 3 ) {
            return signature;
        }

        String variableSignature = split[ 0 ];
        String calledMethod = split[ 1 ];
        String passedType = split[ split.length - 1 ];

        String cleanedVariableSignature = variableSignature.substring( 2 );
        Collection<? extends PhpNamedElement> bySignature = phpIndex.getBySignature( cleanedVariableSignature );
        if ( bySignature.size() == 0 ) {
            return signature;
        }

        FunctionImpl firstSignatureMatch = (FunctionImpl)bySignature.iterator().next();
        return "#M#C" + firstSignatureMatch.getType() +  ":" + calledMethod + ":" + passedType ;
    }
}

File: src/com/ptby/dynamicreturntypeplugin/index/ReturnInitialisedSignatureConverter.kt
package com.ptby.dynamicreturntypeplugin.index

import com.intellij.openapi.project.Project
import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.psi.elements.impl.FunctionImpl
import com.ptby.dynamicreturntypeplugin.signature_extension.withMethodCallPrefix
import com.ptby.dynamicreturntypeplugin.signatureconversion.CustomMethodCallSignature

class ReturnInitialisedSignatureConverter {


    fun convertSignatureToClassSignature(signature: CustomMethodCallSignature,
                                                project: Project): CustomMethodCallSignature {
        val phpIndex = PhpIndex.getInstance(project)

        val cleanedVariableSignature = signature.className.substring(2)
        val bySignature = phpIndex.getBySignature(cleanedVariableSignature)
        if (bySignature.size == 0) {
            return signature
        }

        val firstSignatureMatch = bySignature.iterator().next() as FunctionImpl
        return CustomMethodCallSignature.new(
                firstSignatureMatch.type.withMethodCallPrefix(),
                signature.method,
                signature.desiredParameter
        )
    }
}